File: java/src/ch1/Ex124.java
package ch1;

import java.util.Date;
import java.util.Random;

public class Ex124 {
    public static void main(String[] args) {
        System.out.println(runtime());
        searchForPrimes(1000, 1020);
        searchForPrimes(10000, 10040);
        searchForPrimes(100000, 100080);
        System.out.println("done");
    }

    public static int square(int n) {
        return n * n;
    }

    public static boolean divides(int a, int b) {
        return 0 == b % a;
    }

    public static boolean prime(int n) {
        return fast_prime(n, 3);
    }

    public static boolean fast_prime(int n, int times) {
        if (times == 0)
            return true;
        else if (fermat_test(n))
            return fast_prime(n, times-1);
        else
            return false;
    }

    @FunctionalInterface
    interface TryFermat { boolean apply(int n); }

    public static boolean fermat_test(int n) {
        TryFermat f = (int a) -> a == new ExpMod(a, n, n);
        return f.apply(1 + random(n-1));
    }

    public static int random(int limit) {
        return new Random().nextInt(limit);
    }

    public static long runtime() {
        return new Date().getTime();
    }

    public static void reportPrime(long elapsedTime) {
        System.out.print(" *** ");
        System.out.println(elapsedTime);;
    }

    public static void reportNonPrime() {
        //do nothing System.out.println("  ");
    }

    public static void startPrimeTest(int n, long startTime) {
        if (prime(n)) {
            reportPrime(runtime() - startTime);
        } else {
            reportNonPrime();
        }
    }

    public static void timedPrimeTest(int n) {
        System.out.println();
        System.out.print(n);
        startPrimeTest(n, runtime());
    }

    public static void searchForPrimes(int start, int end) {
        searchHelper(odd(start) ? start : ++start, end);
        System.out.println("finished");
    }

    public static void searchHelper(int n, int end) {
        if (n <= end) {
            timedPrimeTest(n);
            searchHelper(n+2, end);
        } else {
            System.out.println(" ");
        }
    }

    static class ExpMod implements Value<Integer> {
        int base;
        int exp;
        int m;
        ExpMod(int base, int exp, int m) {
            this.base = base;
            this.exp = exp;
            this.m = m;
        }

        @Override
        public Integer value() {
            return calc(this.base, this.exp, this.m);
        }

        int calc(int base, int exp, int m) {
            if (exp == 0) {
                return 1;
            } else if (divides(2, exp)) {
                return square(calc(base, exp/2, m)) % m;
            } else {
                return (base * calc(base, exp-1, m)) % m;
            }
        }
    }
}

File: scala/src/ch1/ex124.scala
// run with> scala exXXX.scala

def divides(a: Int, b: Int): Boolean = b % a == 0

def smallest_divisor(n: Int): Int = find_divisor(n, 2)

def find_divisor(n: Int, test: Int): Int = test match {
  case x if x * x > n => n
  case x if divides(test, n) => test
  case _ => find_divisor(n, test + 1)
}

//def prime(n: Int) = n == smallest_divisor(n)
def prime(n: Int) = n == fast_prime(n, 3)

def fast_prime(n: Long, times: Int): Boolean = times match {
  case 0 => true
  case _ => fermat_test(n) ? fast_prime(n, times-1) : false
}

def fermat_test(n: Int) : Boolean = {
  val try_it = (a: Int) => expmod(a, n, n) == a
  try_it(1 + random(n-1))
}

def expmod(base: Int, exp: Int, m: Int): Int = exp match {
  case 0 => 1
  case _ if divides(2, exp) => (square(expmod(base, exp/2, m)) % m).asInstanceOf(Int)
  case _ => (base * expmod(base, exp-1, m)) % m
}

def square(n: Int): Long = n.asInstanceOf(Long) * n.asInstanceOf(Long)

def random(limit: Int): Int = {
  new Random().nextInt(limit)
}

def report_prime(n: Int) = {
  if (prime(n)) {
    println(n + " ***")
  } else {
    println(n)
  }
}

def search_helper(n: Int, finish: Int): Int = {
  if (n <= finish) {
    report_prime(n)
    search_helper(n + 2, finish)
  } else {
    println("finished\n")
    0
  }
}

def search_for_primes(start: Int, finish: Int) =
  search_helper(if (divides(2, start)) start + 1 else start, finish)


// testing
search_for_primes(1000, 1020)
search_for_primes(10000, 10040)
search_for_primes(100000, 100080)

println("done")
